content/40.units_and_quantities.md

```diff
@@ -1,269 +1,682 @@
+code units, they should also be defined in subclasses by extending this function.
 
-Every `Dataset` has a `length_unit`, `time_unit`, and `mass_unit`, attribute that the user can quickly and easily query to discover the base units of the simulation. For example:
+When code units are set up for a `Dataset`, `unyt_quantity` instances corresponding to these code units are also set up and added as attributes to the `Dataset`. Examples are `length_unit`, `time_unit`, `mass_unit`, and `temperature_unit`:
 
 ```python
 >>> import yt
 >>> ds = yt.load("Enzo_64/DD0043/data0043")
 >>> print(ds.length_unit)
 128 Mpccm/h
->>> print(ds.quan(1.0, "code_length").in_units("Mpccm/h"))
-128 Mpccm/h
 >>> print(ds.length_unit.in_cgs())
-5.55517285026e+26 cm
+5.555172850264421e+26 cm
 ```
 
-Optionally `velocity_unit`, `pressure_unit`, `temperature_unit`, and `density_unit` may be defined as well if the units for these fields cannot be inferred from the mass, length, and time units.
+Optionally, `velocity_unit`, `density_unit`, `pressure_unit`, and `magnetic_unit` may be defined as well,if the units for these fields cannot be inferred from the mass, length, and time units.
 
-Additionally, we allow conversions to the simulation unit system.
-Data in code units are available by converting to `code_length`, `code_mass`, `code_time`, `code_velocity`, `code_density`, `code_magnetic`, `code_pressure`, `code_metallicity`, or any combination of those units.
-Code units preserve dimensionality: an array or quantity that has units of `cm` will be convertible to `code_length`, but not to `code_mass`.
+#### Handling cosmological units {#sec:handling-cosmological-units}
 
-On-disk data are also be available to the user, presented in unconverted code units.
-To obtain on-disk data, a user need only query a data object using an on-disk field name:
+A special case of astrophysical hydrodynamical simulations are cosmological simulations, which
+attempt to simulate the evolution of structure on cosmological length and time scales. Such
+simulations use "comoving coordinates", which is a notion of length that does not change with the
+expansion of the universe, in contrast to normal physical distances which do. 
+
+If we detect that we are loading a cosmological simulation performed in comoving coordinates, extra
+comoving units are added to the dataset's unit registry. Comoving length unit symbols are still
+named following the pattern `<length symbol>cm`, i.e. `Mpccm`. The $h$ symbol is treated as a base
+unit, `h`, which defaults to unity. The `Dataset` updates the `h` symbol to the correct value when
+loading a cosmological simulation.
+
+The change of physical length is illustrated by the following example, of two datasets from the same
+simulation but at different times/scale factors/redshifts:
 
 ```python
 >>> import yt
->>> ds = yt.load("Enzo_64/DD0043/data0043")
+>>> ds1 = yt.load('Enzo_64/DD0002/data0002')
```

Comment:
this is where bug https://github.com/yt-project/yt/issues/4622 was discovered, right ?

Reply:
Yes, but I removed the offending code from this paper draft--I don't feel it's crucial to highlight the feature and I also didn't want to hold up finishing the paper before it was fixed.